falcon/asgi/app.py

```diff
@@ -698,9 +698,9 @@ async def watch_disconnect():
                         '__aiter__ method. Error raised while iterating over '
                         'Response.stream: ' + str(ex)
                     )
-
-            if hasattr(stream, 'close'):
-                await stream.close()
+                finally:
+                    if hasattr(stream, 'close'):
```

Comment:
We shouldn't move `.close()` in here. This way, `.close()` is only attempted for the second branch of the main conditional statement (whether `stream` has `read()` or not). That's why the tests are failing too.

Reply:
Thanks, this makes a lot of sense!